daemon/cmd/daemon_privileged_test.go

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+//go:build privileged_tests
+
+package cmd
+
+import (
+	"net"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/vishvananda/netlink"
+
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/netns"
+)
+
+func Test_removeOldRouterState(t *testing.T) {
+	const netnsName = "test-daemon-priv-0"
+	clean := setupNetNS(t, netnsName)
+	defer clean()
+	defer netns.RemoveNetNSWithName(netnsName)
+
+	// Assert that the old router IP (192.0.2.1) was removed because we are
+	// restoring a different one (10.0.0.1).
+	assert.NoError(t, removeOldRouterState(false, net.ParseIP("10.0.0.1")))
+	addrs, err := netlink.AddrList(&netlink.Dummy{
+		LinkAttrs: netlink.LinkAttrs{
+			Name: defaults.HostDevice,
+		},
+	}, netlink.FAMILY_V4)
+	assert.NoError(t, err)
+	assert.Len(t, addrs, 0)
+
+	// Assert no errors in the case we have no IPs to remove from cilium_host.
+	assert.NoError(t, removeOldRouterState(false, nil))
+
+	// Remove the cilium_host device and assert no error on "link not found"
+	// error.
+	clean()
+	assert.NoError(t, removeOldRouterState(false, nil))
+}
+
+func setupNetNS(t *testing.T, netnsName string) func() {
+	t.Helper()
+
+	dummy := &netlink.Dummy{
+		LinkAttrs: netlink.LinkAttrs{
+			Name: defaults.HostDevice,
+		},
+	}
+	assert.NoError(t, netlink.LinkAdd(dummy))
+	cleanup1 := func() { netlink.LinkDel(dummy) }
+	cleanup := cleanup1
+
+	_, ipnet, _ := net.ParseCIDR("192.0.2.1/32")
+	assert.NoError(t, netlink.AddrAdd(dummy, &netlink.Addr{IPNet: ipnet}))
+
+	netns0, err := netns.ReplaceNetNSWithName(netnsName)
+	assert.NoError(t, err)
+	assert.NotNil(t, netns0)
+	cleanup = func() {
+		cleanup1()
+		netns0.Close()
```

Comment:
This way, in the cleanup, the deletion of the link device will be completed before the cleanup of the network namespace. Is this done on purpose (or is this irrelevant)?

Reply:
Yes I intended for the device to be deleted and then the entire namespace to be deleted. Is that a concern?

Reply:
Nope, that's fine! :+1: